### PR TITLE
Add overlays mechanism to Nixpkgs.

### PR DESCRIPTION
**Pull request by nbp**
This patch add a new argument to Nixpkgs default expression named "overlays".

By default, the value of the argument is either taken from the environment variable `NIXPKGS_OVERLAYS`,
or from the directory `~/.nixpkgs/overlays/`.  If the environment variable does not name a valid directory
then this mechanism would fallback on the home directory.  If the home directory does not exists it will
fallback on an empty list of overlays.

The overlays directory should contain the list of extra Nixpkgs stages which would be used to extend the
content of Nixpkgs, with additional set of packages.  The overlays, i-e directory, files, symbolic links
are used in alphabetical order.

The simplest overlay which extends Nixpkgs with nothing looks like:

```nix
self: super: {
}
```

More refined overlays can use `super` as the basis for building new packages, and `self` as a way to query
the final result of the fix-point.

An example of overlay which extends Nixpkgs with a small set of packages can be found at:
  https://github.com/nbp/nixpkgs-mozilla/blob/nixpkgs-overlay/moz-overlay.nix

To use this file, checkout the repository and add a symbolic link to
the `moz-overlay.nix` file in `~/.nixpkgs/overlays` directory.

###### Motivation

The goal of this patch is to provide an alternative to the `pkgs.overridePackages` function (#20927), which re-execute Nixpkgs fix-point with new inputs.  Instead, the alternative would be use Nixpkgs top-level function, with an `overlays` attribute which would be a list with the same argument as what is today the argument given to `pkgs.overridePackages`.

Thus, in addition to provide a useful extensible layer to Nixpkgs third parties, such as Mozilla, or any other company.  This would remove the existing leakage of Nixpkgs fix-point out of Nixpkgs.  This feature would also be nicer for the `security-updates` branch (#10851), as the overlays would be under the fix-point on which the patching mechanism works on.

###### Things done

- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`

---

### CODE REVIEW

**Comment by Ericson2314** [comment]
Hmm I'll need to think more about the ramifications for security updates, but otherwise I'm confused by the motivation for this. To me, `mkExtensible` is one of the best parts of the fix-extends ideom, and one I hope to see used *more* widely, not less.

CC @cstrahan, who created that abstraction.

**Comment by nbp** [comment]
@Ericson2314, @cstrahan:

The `security-updates` branch is using the fixed function a first time within the fix-point for base packages, and then as a standalone function for patching packages, and as a guide for another fix-point for propagating the patches.

The `makeExtensibleWithCustomName` function create a new attribute which let you inject phases in the pipeline of fixed function.  If we wanted to carry the `security-updates` mechanism for the `pkgs.overridePackages`, then we would have to add special cases for this function, which I consider to be more painful than useful.

By adding the `overlays` list of extra phases, you can easily combine orthogonal extensions of Nixpkgs, which surpass the `pkgs.overridePackages` which only take one extra phase. Moreover, the  `pkgs.overridePackages` function can easily be emulated as demonstrated in https://github.com/NixOS/nixpkgs/pull/20927#issuecomment-267474647

Thus, I see no need to keep the `pkgs.overridePackages` function.

File: doc/overlays.xml
Comment (the-kenny) [inline doc/overlays.xml]: How does this option work with different users? Will only the overlay set of the user executing `nixos-rebuild` considered or will it always use root's?

File: doc/overlays.xml
Comment (the-kenny) [inline doc/overlays.xml]: Is it possible to hard-code this directory for `nixos-rebuild` from `configuration.nix`? I think that might be useful.

**Review by the-kenny — COMMENTED** [review]
Some directory-resolution related questions.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: With the current patch, it will take the overlays of the user which is executing the command.

With the above suggestion, none should be considered unless they are specified in the `configuration.nix` file, or any other module.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: I agree.

This would be a terrible experience if one `configuration.nix` is not evaluated the same way across computer depending on the `~/.nixpkgs/overlays` directory content.

I will add the `overlays` argument to `nixos/modules/misc/nixpkgs.nix`.

File: doc/overlays.xml
Comment (teh) [inline doc/overlays.xml]: "phases" sounds like pre or post-phase. Is a "phase" one file (and one function), and executed in alphabetical order?

File: doc/overlays.xml
Comment (teh) [inline doc/overlays.xml]: Why woulf we ignore invalid ones instead of throwing an error? This seems like a recipe for "silent, hard-to-debug" problems?

File: doc/overlays.xml
Comment (teh) [inline doc/overlays.xml]: Does `NIXPKGS_OVERLAYS` support colon separated paths like `PATH`?

File: nixos/doc/manual/release-notes/rl-1703.xml
Comment (teh) [inline nixos/doc/manual/release-notes/rl-1703.xml]: That's a very long space after "exists." Maybe a missing line break?

**Review by teh — APPROVED** [review]
This is really amazing work & a lot of power for such a small change! I'm very much in favour.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: Yes, one phase / layer is one overlay, which is a single function `self: super: { ... }`. Maybe I should use the term layer instead of phase?

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: What I meant is the fallback logic, we consider the environment variable `NIXPKGS_OVERLAYS`, and check if this is a directory.  Then we do the same check for the `$HOME/.nixpkgs/overlays`.

I agree, we should throw if `NIXPKGS_OVERLAYS` is set, but the content does not validate.
But I do not think we should do anything if the `$HOME/.nixpkgs/overlays` directory does not exists.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: For the moment no, I think this would be a nice addition in a future patch.

I also think being able to pull a Nix expression from a trusted remote (such as `https://my.company/overlay.nix`) would also be interesting.

File: doc/overlays.xml
Comment (acowley) [inline doc/overlays.xml]: I think "layer" would be better, too.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: This is fixed in commit 46ae6b5

**Comment by Ericson2314** [comment]
Note in https://github.com/NixOS/nixpkgs/pull/21268/commits/22784143012662f14db34b03fd073f4d8f69ab9c#diff-19c9515018ff76dea736b57000a62d5bR49 I use `overridePackages` in a way that I worry would be hard to replicate with this.

**Comment by Ericson2314** [comment]
More broadly, as I wrote in https://github.com/NixOS/nixpkgs/pull/16531#issuecomment-265489603 I think I thinking instead of having nixpkgs vs nixos, we should have nixpkgs vs "nixuser" vs "nixos". nixkpgs would have no config attrset, and no `impure.nix`, and basically just be packages. nixuser would have some sort of config system (probably module system), things like firefoxWrapper, dotfiles management if we ever get that.

I think splitting the pure and impure parts of this would be an excellent further justification for factoring out this "nixuser".

File: nixos/doc/manual/release-notes/rl-1703.xml
Comment (domenkozar) [inline nixos/doc/manual/release-notes/rl-1703.xml]: There should be a link to the manual with docs.

File: nixos/doc/manual/release-notes/rl-1703.xml
Comment (domenkozar) [inline nixos/doc/manual/release-notes/rl-1703.xml]: This is not clear what `pkgs` is and what would be the contents of `overlays`. It's better to point to overlays section in the manual.

**Review by domenkozar — DISMISSED** [review]
Overall looks good.

I think documentation needs someone who speaks English natively to flesh it out.

For Overlays I think we should have a tool that manages them. Something similar to git submodule that would sync git repositories. We should decide to go this route from the first day, prohibiting manipulation of `~/.nixpkgs/overlays` by hand.

**Comment by teh** [comment]
I gave this a spin and I'm not super convinced that a directory is the best public API (it's not terrible either). The additional `overlays` parameter however is awesome!

@domenkozar I think the docs are OK - do you have a specific thing in mind? Also, what kind of tool do you have in mind?

Both @nbp & @domenkozar - what do you think about splitting the filesystem API (`~/.nixpkgs/overlays` and `NIXPKGS_OVERLAYS`) into a separate branch so we can get the uncontroversial parts merged soon? I can do the work :)

File: nixos/doc/manual/release-notes/rl-1703.xml
Comment (nbp) [inline nixos/doc/manual/release-notes/rl-1703.xml]: @domenkozar How would you link to the new Nixpkgs documentation from the release notes?

**Comment by nbp** [comment]
I think being able to manage overlays by hand is a nice option for experts, probably not the mainstream one to advertise, I agree.

@teh, The file system API is the one which makes the overlays accessible to a large set of users, especially with a tool to manipulate these overlays from a command line.

@domenkozar, Do you have any native English speaker in mind who could do the review of the documentation?

**Comment by nbp** [comment]
@Ericson2314 
> Note in 2278414#diff-19c9515018ff76dea736b57000a62d5bR49 I use overridePackages in a way that I worry would be hard to replicate with this.

Let me disagree, this can easily be replicated by exposing under nixpkgs the set of overlays which are being used, such that you can extend it in the next Nixpkgs import.

The comment above `overridePackages` states that it should not be used in Nixpkgs, and I would recommend no less for overlays, as in both cases we are re-evaluating Nixpkgs fix-point.  To this point I would also agree that we already do that in a different way for stdenv and consumes 60 MB in all Nixpkgs evaluation.  Changing from one way of consuming 60 MB to another way is point-less.

This 60 MB consumption is a problem for another bug, which is that stdenv bootstrapping should not use `allPackages` function to create a new fix-point, but re-use the current one to extract the recipes of the packages that it needs for the bootstrap.

**Comment by Ericson2314** [comment]
> The comment above overridePackages states that it should not be used in Nixpkgs, and I would recommend no less for overlays, as in both cases we are re-evaluating Nixpkgs fix-point.

I copied that note. I am not at all sure the extends idiom is expensive as the original fix-point is only "barely evaluated".

> This 60 MB consumption is a problem for another bug, which is that stdenv bootstrapping should not use allPackages function to create a new fix-point, but re-use the current one to extract the recipes of the packages that it needs for the bootstrap.

Ok this is a much larger conversation. I'm on the fence about whether this is a good idea for the existing bootstrapping, but it definitely isn't a replacement for cross which as far as I can tell requires multiple fixes. Let's talk about this in my cross PR.

**Comment by nbp** [comment]
I think this changes are good, and should be merged, and if possible before they bit-rot once more.
I fixed the documentation issue reported by @domenkozar , but I have not yet made a tool to handle repository under `~/.nixpkgs/overlay`, because I do not think this is yet necessary to enforce such thing yet.

I think having a tool would be better once this feature is getting more adoptions from users and companies.  Then we can standardize what layout of overlay repositories should be.

Concerning having a native speaker reviewing the documentation, I did not got any answer from @domenkozar.  Maybe @zimbatm could do the review of the documentation?

**Comment by domenkozar** [comment]
Sounds good to me. I think tooling would leverage this code, but I understand you have even less time than me :) So if we can get @zimbatm to review docs, this sounds like a good step forward.

**Comment by aneeshusa** [comment]
I'm a native English speaker, I can try to review the docs later today.

**Review by fpletz — APPROVED** [review]

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: grammar: `As an argument ...`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: grammar: `which is described in`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: `In the directory pointed to by the environment variable <varname>NIXPKGS_OVERLAYS</varname>.`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: `In the directory <filename>~/.nixpkgs/overlays/</filename>.`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: `..., the directory should contain Nix expressions defining the overlays. Each overlay can be a file, a directory containing a default.nix file, or a symlink to one of those. The expressions should follow the syntax described in ...`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: grammar: `To install an overlay using the last option, you can clone the overlay's repository and add a symbolic link to it in the ...`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: Overlays can also be used for adding packages, so I would make this a bit more general, maybe `used by Nixpkgs to compose the set of all packages`.

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: No need to mention symlinks here since it's described below.

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: No need to mention symlinks here since it's described below.

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: Also, this section should describe any ordering guarantees (or lack thereof) about overlays retrieved from a directory.

**Review by aneeshusa — CHANGES_REQUESTED** [review]
Overall docs look good, this is mostly small grammar fixes.

Also a request for a special `callPackage` to make using overlays easier.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: I guess we can omit this part, as we can always add a new attribute name to get any `super` value into `self`.  So I think I will only keep the `recipes` part of this sentence and remove the `dependency` part.

**Comment by nbp** [comment]
@aneeshusa , Thanks a lot of the review, I will fix the nits tomorrow.

For the custom `callPackage`, I think that is a good idea, I will think about it and the consequences/applications with the security-updates branch.

**Comment by aneeshusa** [comment]
FYI, just noticed I had made a typo in my example for a custom `callPackage`, I fixed it at https://github.com/NixOS/nixpkgs/pull/21243#discussion_r96124098.

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: I added the following paragraph right after to describe the ordering of overlays, which has an impact both when providing a list of `overlays` as arguments of Nixpkgs, or when using a directory.

`The order of the overlay layers can influence the recipe of packages if two layers override
the same recipe. In the case where overlays are loaded from a directory, these are loaded in
alphabetical order.`

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: Looking at the current source code, you can already do that by using `super.callPackage`, which is already defined to take the dependencies from `self`.  I will update the example accordingly, and add additional overridden inputs, to highlight usage of `self` in opposition to `super`.

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: grammar: options

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: `they are loaded`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: `link to it in`

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: Needs a period or colon at the end.

File: doc/overlays.xml
Comment (aneeshusa) [inline doc/overlays.xml]: Why is this `super.callPackage`? I would think `super.callPackage` would use packages from `super` for dependencies, where the documentation says to use packages from `self`.

Comment (aneeshusa) [inline doc/overlays.xml]: E.g. I would think we need to provide a custom `callPackage` for each `self` layer using
`lib.callPackagesWith`.

**Review by aneeshusa — CHANGES_REQUESTED** [review]
A few more grammar nits, and some confusion about the usage of `callPackage` here.

IDK why GitHub is showing these as outdated, they should still be valid comments at time of writing...

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]: Technically you could use both, but this would get some extra meaning in the futur (security-update branch), because `self.callPackage` goes twice through the `self` attribute evaluation.  The first one to resolve the `callPackage` function, and the second time to resolve the dependencies from `self` which are captured by the `callPackage` function.

Note that `self` is the same provided to all layers.  So using `self.callPackage` is for the moment identical to `super.callPackage`, except for this extra hop through the fix-point.

I documented it as functions should be taken from `super`, because they already capture `self` if they have to.

Also note that this way, we make it syntactically verifiable, as anything which comes around the dependencies uses `super`, and anything which is use to find dependencies uses `self`.

Comment (aneeshusa) [inline doc/overlays.xml]: That sounds reasonable, but can you explain how functions already capture `self` if they need to? E.g. if I have an overlay that adds two new packages:
```
self: super: {
    # Let's say all deps are already in super,
    # so I can see this working with super.callPackages or self.callPackages
    foo = super.callPackage {};
    # But this depends on foo, which isn't in super
    bar = super.callPackage {}; 
}
```
How does the second `super.callPackage` get a reference to `foo` to provide to `bar`?

Comment (aneeshusa) [inline doc/overlays.xml]: OK, I tried it out and looked at the implementation and I now understand why this works, albeit the current behavior is slightly surprising to me.

**Review by aneeshusa — APPROVED** [review]

File: doc/overlays.xml
Comment (nbp) [inline doc/overlays.xml]:
```nix
self: super:

{
  callWithSelf = f: f self;
}
```

Any layer defined after could then do:

```nix
self: super:

{
  mySelfClone = super.callWithSelf (x: x);
}
```

**Comment by nbp** [comment]
Tested locally on top of nixpkgs-unstable channel:
 - [ ] `./maintainers/scripts/travis-nox-review-pr.sh nixpkgs-verify nixpkgs-manual nixpkgs-tarball nixpkgs-unstable`
- [ ] `./maintainers/scripts/travis-nox-review-pr.sh nixos-options nixos-manual`

File: nixos/doc/manual/release-notes/rl-1703.xml
Comment (teh) [inline nixos/doc/manual/release-notes/rl-1703.xml]: just noticed a double `in` (see next line) - @nbp if you have direct push permissions maybe a quick fix? Otherwise I'll try to mop it up with some other change.

**Comment by Ericson2314** [comment]
Wait, do overlays currently apply to bootstrapping stages after all?

File: pkgs/top-level/default.nix
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: If we do want all bootstraping stages to include the overlay unconditionally, we could instead "partially apply" overlays to all packages (3 lines above the `inherit lib nixpkgsFun;`, but I cannot comment there because unchanged).

**Comment by nbp** [comment]
@Ericson2314 
> Wait, do overlays currently apply to bootstrapping stages after all?

Yeah, but I think it would makes sense to disable that as soon as we can move the bootstraping process in Nixpkgs fix-point, as this would no longer be needed.  For the moment I would not recommend anybody to rely on overlays for changing the stdenv, except by doing it explicitly.

@teh Thanks, I will fix it later, if nobody does it before ;)

**Comment by Ericson2314** [comment]
@nbp, well, to be clear, we can disable it in the meantime too. But that plan sounds fine to me.

File: nixos/doc/manual/release-notes/rl-1703.xml
Comment (nbp) [inline nixos/doc/manual/release-notes/rl-1703.xml]: This is fixed as part of commit 0214d94. Thanks!

**Comment by edolstra** [comment]
IMHO we should not add more ad hoc environment variables like $NIXPKGS_OVERLAYS. Why not use the NIX_PATH mechanism for this?

**Comment by edolstra** [comment]
I've changed `$NIXPKGS_OVERLAYS` to `<nixpkgs-overlays>`, and `~/.nixpkgs/overlays` to `~/.config/nixpkgs/overlays` in order to not add more non-XDG-compliant paths.

BTW, since the point of overlays is that they can be stacked easily, should the overlays in `<nixpkgs-overlays>` and `~/.config/nixpkgs/overlays` be merged, rather than having the former override the latter?

**Comment by nbp** [comment]
@edolstra 
> BTW, since the point of oveprlays is that they can be stacked easily, should the overlays in `<nixpkgs-overlays>` and `~/.config/nixpkgs/overlays` be merged, rather than having the former override the latter?

I honestly do not know yet what would be the most desirable behaviour.

My initial though was that we could replace the default configuration and test with a different one, which is interesting in terms of reproducibility.

**Pull request merged**
